File: scripts/install_project_skills.py
#!/usr/bin/env python3
"""把仓库中的 Unity 协作 Skills 复制到目标项目。"""

from __future__ import annotations

import argparse
import shutil
import sys
from pathlib import Path


REPOSITORY_ROOT = Path(__file__).resolve().parents[1]


def discover_skill_directories() -> tuple[Path, ...]:
    """按名称发现仓库根目录中的可安装 Skill，避免维护重复清单。"""
    return tuple(
        path
        for path in sorted(REPOSITORY_ROOT.iterdir(), key=lambda item: item.name)
        if path.is_dir() and (path / "SKILL.md").is_file()
    )


def validate_project_root(project_root: Path) -> Path:
    """要求目标是已有项目目录，并拒绝文件系统根目录。"""
    resolved = project_root.resolve()
    if not resolved.is_dir():
        raise ValueError(f"目标项目目录不存在或不是目录：{resolved}")
    if resolved == resolved.parent:
        raise ValueError("目标项目不能是文件系统根目录。")
    return resolved


def reject_symlink(path: Path, label: str) -> None:
    """拒绝符号链接目标，避免覆盖授权沿链接逃逸到项目目录之外。"""
    if path.is_symlink():
        raise ValueError(f"{label}不能是符号链接：{path}")


def install_skills(project_root: Path, force: bool) -> list[Path]:
    """复制全部 Skill；只有显式 force 才替换目标中的同名目录。"""
    target_root = project_root / ".agents" / "skills"
    reject_symlink(project_root / ".agents", ".agents 目录")
    reject_symlink(target_root, "Skills 目录")
    sources = discover_skill_directories()
    targets = [target_root / source.name for source in sources]
    for target in targets:
        reject_symlink(target, "Skill 目标")
    existing = [target for target in targets if target.exists()]
    if existing and not force:
        names = "、".join(target.name for target in existing)
        raise FileExistsError(f"拒绝覆盖已有 Skills：{names}。如确需替换，请传入 --force。")

    target_root.mkdir(parents=True, exist_ok=True)
    for source, target in zip(sources, targets, strict=True):
        if target.exists():
            # --force 是用户显式授权；删除范围被限制在 .agents/skills 下的已发现同名目录。
            shutil.rmtree(target)
        shutil.copytree(source, target)
    return targets


def parse_args() -> argparse.Namespace:
    """解析目标项目和显式覆盖授权。"""
    parser = argparse.ArgumentParser(description="安装 Unity 游戏协作 Skills")
    parser.add_argument("--project-root", required=True, type=Path, help="目标项目根目录")
    parser.add_argument("--force", action="store_true", help="替换目标中的同名 Skills")
    return parser.parse_args()


def main() -> int:
    """执行安装并输出每个可审计目标路径。"""
    args = parse_args()
    try:
        project_root = validate_project_root(args.project_root)
        installed = install_skills(project_root, args.force)
    except (OSError, ValueError) as error:
        print(f"安装失败：{error}", file=sys.stderr)
        return 1

    print("已安装 Unity 游戏协作 Skills：")
    for path in installed:
        print(path)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/install-project-skills.mjs
#!/usr/bin/env node

import { cpSync, existsSync, lstatSync, mkdirSync, rmSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const PACKAGE_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const SKILL_NAMES = Object.freeze([
  "unity-development-workflow",
  "unity-game-architecture",
  "unity-game-audio",
  "unity-game-balance",
  "unity-game-production",
  "unity-game-qa-performance",
  "unity-game-release",
  "unity-game-visual-assets",
  "unity-gameplay-development",
]);

/**
 * 解析目标目录和显式覆盖授权。
 *
 * @returns {{ help: boolean, force: boolean, requestedPath: string | undefined }} 安装参数。
 */
function parseArguments() {
  let help = false;
  let force = false;
  let requestedPath;

  for (const argument of process.argv.slice(2)) {
    if (argument === "--help" || argument === "-h") {
      help = true;
    } else if (argument === "--force") {
      force = true;
    } else if (argument.startsWith("-")) {
      throw new Error("未知参数：" + argument);
    } else if (requestedPath === undefined) {
      requestedPath = argument;
    } else {
      throw new Error("最多只能传入一个目标项目目录。");
    }
  }

  return { help, force, requestedPath };
}

/**
 * 解析目标项目目录，并拒绝不存在、非目录或文件系统根目录。
 *
 * @param {string | undefined} requestedPath 用户传入的目标目录。
 * @returns {string} 规范化后的绝对目录。
 */
function parseTargetDirectory(requestedPath) {
  const targetDirectory = resolve(requestedPath ?? process.cwd());

  if (!existsSync(targetDirectory) || !statSync(targetDirectory).isDirectory()) {
    throw new Error("目标项目目录不存在或不是目录：" + targetDirectory);
  }
  rejectSymbolicLink(targetDirectory, "目标项目目录");
  if (dirname(targetDirectory) === targetDirectory) {
    throw new Error("目标项目不能是文件系统根目录：" + targetDirectory);
  }

  return targetDirectory;
}

/**
 * 拒绝符号链接或 Windows Junction，避免覆盖操作越过目标项目边界。
 *
 * @param {string} path 待检查路径。
 * @param {string} label 错误信息中的路径标签。
 */
function rejectSymbolicLink(path, label) {
  if (existsSync(path) && lstatSync(path).isSymbolicLink()) {
    throw new Error(label + "不能是符号链接或目录联接：" + path);
  }
}

/**
 * 校验 npm 包中固定的九个 Skill 完整存在。
 *
 * @returns {Array<{ name: string, source: string }>} 已验证的安装源。
 */
function validateSources() {
  return SKILL_NAMES.map((name) => {
    const source = resolve(PACKAGE_ROOT, name);
    if (!existsSync(source) || !statSync(source).isDirectory()) {
      throw new Error("安装包缺少 Skill 目录：" + name);
    }
    if (!existsSync(resolve(source, "SKILL.md"))) {
      throw new Error("安装包缺少 Skill 入口：" + name + "/SKILL.md");
    }
    return { name, source };
  });
}

/**
 * 把当前 npx 包内的全部 Skills 复制到目标项目。
 *
 * @param {string} projectRoot 已验证的项目根目录。
 * @param {boolean} force 是否显式替换同名 Skill。
 * @returns {string[]} 安装后的绝对路径。
 */
function installSkills(projectRoot, force) {
  const sources = validateSources();
  const agentsRoot = resolve(projectRoot, ".agents");
  const targetRoot = resolve(agentsRoot, "skills");
  const targets = sources.map(({ name, source }) => ({ name, source, target: resolve(targetRoot, name) }));

  rejectSymbolicLink(agentsRoot, ".agents 目录");
  rejectSymbolicLink(targetRoot, "Skills 目录");
  for (const { target } of targets) {
    rejectSymbolicLink(target, "Skill 目标");
  }

  const existing = targets.filter(({ target }) => existsSync(target));
  if (existing.length > 0 && !force) {
    const names = existing.map(({ name }) => name).join("、");
    throw new Error("拒绝覆盖已有 Skills：" + names + "。如确需替换，请传入 --force。");
  }

  mkdirSync(targetRoot, { recursive: true });
  for (const { source, target } of targets) {
    if (existsSync(target)) {
      // --force 是用户显式授权；删除范围只限固定白名单中的同名 Skill 目录。
      rmSync(target, { recursive: true, force: true });
    }
    cpSync(source, target, { recursive: true, errorOnExist: true });
  }

  return targets.map(({ target }) => target);
}

/**
 * 执行安装并输出可审计的目标路径。
 */
function main() {
  const { help, force, requestedPath } = parseArguments();
  if (help) {
    console.log("用法：npx -y github:weberwang/unity-skills [目标项目目录] [--force]");
    return;
  }

  const projectRoot = parseTargetDirectory(requestedPath);
  const installed = installSkills(projectRoot, force);
  console.log("已安装 Unity 游戏协作 Skills：");
  for (const path of installed) {
    console.log(path);
  }
}

try {
  main();
} catch (error) {
  console.error("安装失败：" + (error instanceof Error ? error.message : String(error)));
  process.exitCode = 1;
}
